ios/browser/api/storekit_receipt/storekit_receipt.mm

```diff
@@ -0,0 +1,419 @@
+#include "third_party/boringssl/src/include/openssl/pool.h"
+#include "third_party/boringssl/src/pki/input.h"
+#include "third_party/boringssl/src/pki/parse_values.h"
+#include "third_party/boringssl/src/pki/parser.h"
+
+// MARK: - Implementation
+
+namespace brave {
+namespace storekit_receipt {
+bssl::UniquePtr<PKCS7> pkcs7_from_data(NSData* data) {
+  const std::uint8_t* der_bytes =
+      static_cast<const std::uint8_t*>([data bytes]);
+  return bssl::UniquePtr<PKCS7>(d2i_PKCS7(nullptr, &der_bytes, [data length]));
+}
+
+bool decode_asn1_boolean(bssl::der::Input value) {
+  std::uint8_t result = 0;
+  bssl::der::Parser parser(value);
+  if (parser.ReadUint8(&result)) {
+    return static_cast<bool>(result);
+  }
+  return false;
+}
+
+NSUInteger decode_asn1_integer(bssl::der::Input value) {
+  std::uint64_t result = 0;
+  bssl::der::Parser parser(value);
+  if (parser.ReadUint64(&result)) {
+    return result;
+  }
+  return 0;
+}
+
+NSString* decode_asn1_string(bssl::der::Input value) {
+  if (value.empty()) {
+    return nullptr;
+  }
+
+  int cls = 0;
+  int type = 0;
+  long length = 0;
+
+  const std::uint8_t* data = value.data();
+  ASN1_get_object(&data, &length, &type, &cls, value.size());
+
+  if (type == V_ASN1_UTF8STRING) {
+    return base::SysUTF8ToNSString(
+        std::string(reinterpret_cast<const char*>(data), length));
+  }
+
+  if (type == V_ASN1_IA5STRING) {
+    std::string result;
+    if (bssl::der::ParseIA5String(bssl::der::Input(base::make_span(
+                                      data, static_cast<std::size_t>(length))),
+                                  &result)) {
+      return base::SysUTF8ToNSString(result);
+    }
+    return nullptr;
+  }
+
+  return base::SysUTF8ToNSString(
+      std::string(reinterpret_cast<const char*>(data), length));
+}
+
+NSDate* decode_asn1_date(bssl::der::Input value) {
+  NSString* date_string = decode_asn1_string(value);
+  if (date_string) {
+    NSISO8601DateFormatter* formatter = [[NSISO8601DateFormatter alloc] init];
+    return [formatter dateFromString:date_string];
+  }
+  return nullptr;
+}
+
+bool pkcs7_get_signed_content(
+    NSData* data,
+    std::vector<bssl::UniquePtr<ASN1_STRING>>& result) {
+  // 1.2.840.113549.1.7.2
+  static const std::uint8_t kPKCS7SignedData[] = {0x2a, 0x86, 0x48, 0x86, 0xf7,
+                                                  0x0d, 0x01, 0x07, 0x02};
+
+  // Initialize cbs - No BER to DER conversion required
+  CBS cbs;
+  CBS_init(&cbs, static_cast<const std::uint8_t*>([data bytes]), [data length]);
+
+  // Parse content info
+  CBS content_info;
+  if (!CBS_get_asn1(&cbs, &content_info, CBS_ASN1_SEQUENCE)) {
+    return false;
+  }
+
+  // Parse content type
+  CBS content_type;
+  if (!CBS_get_asn1(&content_info, &content_type, CBS_ASN1_OBJECT)) {
+    return false;
+  }
+
+  // We only care about signed data
+  if (!CBS_mem_equal(&content_type, kPKCS7SignedData,
+                     sizeof(kPKCS7SignedData))) {
+    return false;
```

Comment:
Correct it shouldn't be false, but I also do not want to crash the app.
Returning false here is fine imo.
If the receipt cannot be parsed, then we should get a nil value on the Swift side for the receipt instance, and throw an error there instead of crashing in C++.